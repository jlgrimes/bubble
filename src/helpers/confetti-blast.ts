import confetti from 'canvas-confetti';

export const smallConfetti = () => {
  confetti({
    particleCount: 100,
    spread: 70,
    origin: { y: 0.6 },
  });
};

export const lotsOfConfetti = () => {
  var count = 200;
  var defaults = {
    origin: { y: 0.7 },
  };

  function fire(particleRatio: number, opts: Object) {
    confetti(
      Object.assign({}, defaults, opts, {
        particleCount: Math.floor(count * particleRatio),
      })
    );
  }

  fire(0.25, {
    spread: 26,
    startVelocity: 55,
  });
  fire(0.2, {
    spread: 60,
  });
  fire(0.35, {
    spread: 100,
    decay: 0.91,
    scalar: 0.8,
  });
  fire(0.1, {
    spread: 120,
    startVelocity: 25,
    decay: 0.92,
    scalar: 1.2,
  });
  fire(0.1, {
    spread: 120,
    startVelocity: 45,
  });
};
