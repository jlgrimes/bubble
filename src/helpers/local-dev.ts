export const isLocal = (): boolean => window.location.href.includes('localhost');