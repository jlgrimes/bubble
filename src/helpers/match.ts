import { MatchResult } from "../types/Match";

/**
 * Makes a win a loss, a loss a win, a tie a tie. Used for converting match to player match.
 */
 export const inverseResult = (result: MatchResult): MatchResult => {
  if (result === 'win') {
    return 'loss';
  }

  if (result === 'loss') {
    return 'win';
  }

  return 'tie';
};