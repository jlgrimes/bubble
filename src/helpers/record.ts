import { Record } from "../types/Record";
import { MatchResult } from "../types/Match";

export const getUpdatedRecordAfterMatch = (record: Record, result: MatchResult): Record => {
  if (result === 'win') {
    return { ...record, wins: record.wins + 1 };
  }

  if (result === 'loss') {
    return { ...record, losses: record.losses + 1 };
  }

  return { ...record, ties: record.ties + 1 };
}