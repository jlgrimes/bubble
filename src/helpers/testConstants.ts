import type { Match } from '../features/Tournament/Pairings/types';
import type { Player } from '../features/Tournament/Player/types';

export const SAMPLE_PLAYER: Player = {
  id: '0',
  name: 'Player',
  matches: [
    { opponentId: '1', result: 'win' },
    { opponentId: '2', result: 'loss' },
    { opponentId: '3', result: 'tie' },
  ],
  record: {
    wins: 1,
    ties: 1,
    losses: 1
  },
  matchPoints: 4
};

export const generateEmptyPlayers = (number: number) => Array.from(Array(number).keys()).map((id) => ({
  id: `${id}`,
  name: `Player ${id}`,
  matches: [],
  record: {
    wins: 0,
    ties: 0,
    losses: 0
  },
  matchPoints: 0
}));

export const SAMPLE_EMPTY_PLAYERS = [1, 2, 3, 4, 5, 6, 7, 8].map((id) => ({
  id: `${id}`,
  name: `Player ${id}`,
  matches: [],
  record: {
    wins: 0,
    ties: 0,
    losses: 0
  },
  matchPoints: 0
}));

export const SAMPLE_MATCH_TIERED_PLAYERS: Player[] = [
  {
    id: '0',
    name: 'Player',
    matches: [],
    record: {
      wins: 2,
      ties: 0,
      losses: 0
    },
    matchPoints: 6
  }, {
    id: '1',
    name: 'Player',
    matches: [],
    record: {
      wins: 2,
      ties: 0,
      losses: 0
    },
    matchPoints: 6
  }, {
    id: '2',
    name: 'Player',
    matches: [],
    record: {
      wins: 1,
      ties: 0,
      losses: 1
    },
    matchPoints: 3
  }, {
    id: '3',
    name: 'Player',
    matches: [],
    record: {
      wins: 1,
      ties: 0,
      losses: 1
    },
    matchPoints: 3
  }, {
    id: '4',
    name: 'Player',
    matches: [],
    record: {
      wins: 1,
      ties: 0,
      losses: 1
    },
    matchPoints: 3
  }, {
    id: '5',
    name: 'Player',
    matches: [],
    record: {
      wins: 1,
      ties: 0,
      losses: 1
    },
    matchPoints: 3
  }, {
    id: '6',
    name: 'Player',
    matches: [],
    record: {
      wins: 0,
      ties: 0,
      losses: 2
    },
    matchPoints: 0
  }, {
    id: '7',
    name: 'Player',
    matches: [],
    record: {
      wins: 0,
      ties: 0,
      losses: 2
    },
    matchPoints: 0
  }
];

export const SAMPLE_ODD_MATCH_TIERED_PLAYERS: Player[] = [
  {
    id: '0',
    name: 'Player',
    matches: [],
    record: {
      wins: 2,
      ties: 0,
      losses: 0
    },
    matchPoints: 6
  },  {
    id: '1',
    name: 'Player',
    matches: [],
    record: {
      wins: 1,
      ties: 0,
      losses: 1
    },
    matchPoints: 3
  }, {
    id: '2',
    name: 'Player',
    matches: [],
    record: {
      wins: 1,
      ties: 0,
      losses: 1
    },
    matchPoints: 3
  }, {
    id: '3',
    name: 'Player',
    matches: [],
    record: {
      wins: 0,
      ties: 0,
      losses: 2
    },
    matchPoints: 0
  }
];

export const JARED: Player = {
  id: '0',
  name: 'Jared',
  matches: [
    { opponentId: '3', result: 'win' },
    { opponentId: '1', result: 'win' },
  ],
  record: {
    wins: 2,
    ties: 0,
    losses: 0
  },
  matchPoints: 6
};

export const NOAH: Player = {
  id: '1',
  name: 'Noah',
  matches: [
    { opponentId: '2', result: 'win' },
    { opponentId: '1', result: 'loss' },
  ],
  record: {
    wins: 1,
    ties: 0,
    losses: 1
  },
  matchPoints: 3
};

export const RYAN: Player = {
  id: '2',
  name: 'Ryan',
  matches: [
    { opponentId: '1', result: 'loss' },
    { opponentId: '3', result: 'win' },
  ],
  record: {
    wins: 1,
    ties: 0,
    losses: 1
  },
  matchPoints: 3
};

export const KENNY: Player = {
  id: '3',
  name: 'Kenny',
  matches: [
    { opponentId: '0', result: 'loss' },
    { opponentId: '0', result: 'loss' },
  ],
  record: {
    wins: 0,
    ties: 0,
    losses: 2
  },
  matchPoints: 0
};

export const SAMPLE_UNSORTED_PLAYER_LIST: Player[] = [
  KENNY, JARED, NOAH, RYAN
];

export const SAMPLE_SORTED_PLAYER_LIST: Player[] = [
  JARED, NOAH, RYAN, KENNY
];

export const SAMPLE_MATCH_RESULTS: Match[] = [
  // Jared beat player 4
  { playerIds: ['4', '0'], result: 'loss' },
  // Noah beat player 5
  { playerIds: ['5', '2'], result: 'loss' },
  // Ryan lost to player 6
  { playerIds: ['6', '2'], result: 'win' },
  // Kenny lost to player 7
  { playerIds: ['3', '7'], result: 'loss' },
]

export const SAMPLE_PAIRINGS: string[][] =  [
  ['0', '1'],
  ['2', '3']
];