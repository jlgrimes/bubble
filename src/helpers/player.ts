import { Match, PlayerMatch } from "../types/Match";
import { Player } from "../types/Player";
import { getMatchPoints, getUpdatedRecordAfterMatch } from "./record";
import { inverseResult } from "./match";
import { shuffle } from "./shuffle";

export const getActivePlayers = (players: Player[]): Player[] => players.filter(player => !player.dropped);

/**
 * Converts the match object to PlayerMatch, which will be stored in the player object.
 */
export const convertMatchToPlayerMatch = (player: Player, match: Match): PlayerMatch => {
  /**
   * If the first player is the player we're talking about, the opponent is the second player, keep the result the same.
   */
  if (player.id === match.playerIds[0] ) {
    return {
      opponentId: match.playerIds[1],
      result: match.result
    };
  }

  /**
   * If the second player is the player we're talking about, the opponent is the first player, and inverse the match result to store.
   */
  return {
    opponentId: match.playerIds[0],
    result: inverseResult(match.result)
  };
};

export const getUpdatedPlayerAfterMatch = (player: Player, match: Match): Player => {
  const matchResult = convertMatchToPlayerMatch(player, match);
  const newRecord = getUpdatedRecordAfterMatch(player.record, matchResult.result);

  return {
    ...player,
    matches: [...player.matches, matchResult],
    record: newRecord,
    matchPoints: getMatchPoints(newRecord)
  }
}

/**
 * Chunks the players array by match points. Needed to scramble match point tiers for pairings.
 * 
 * @param players Players array
 * @returns Chunked array of player indexes from original player array.
 */
 export const chunkSortedArrayByMatchPoints = (players: Player[]): number[][] => {
  if (players.length === 0) {
    return [];
  }

  // Stores a list of player ids to be mapped to players
  let resultingIdArray: number[][] = [];
  // Stores player ids
  let chunk: number[] = [];
  let chunkMatchPoints: number = players[0].matchPoints;

  for (let playerIdx = 0; playerIdx < players.length; playerIdx++) {
    const player = players[playerIdx];

    if (player.matchPoints === chunkMatchPoints) {
      chunk.push(playerIdx);
    } else {
      resultingIdArray.push(chunk);

      // if we're at the end of the players array and the match point is different, push it as its own entity.
      if (playerIdx === players.length - 1) {
        resultingIdArray.push([playerIdx])
      } else {
        chunkMatchPoints = player.matchPoints;
        chunk = [playerIdx];
      }
    }
  }

  return resultingIdArray;
};

/**
 * Gets updated players for both players in a match after match results are completed.
 * 
 * @param match 
 * @param players 
 */
 export const getUpdatedPlayerPairAfterMatch = (match: Match, players: Player[]): Player[] => {
  const firstPlayer: Player | undefined = players.find((player) => player.id === match.playerIds[0]);
  const secondPlayer: Player | undefined = players.find((player) => player.id === match.playerIds[1]);

  let playerPair: Player[] = [];

  if (firstPlayer) {
    playerPair = [...playerPair, getUpdatedPlayerAfterMatch(firstPlayer, match)]
  }

  if (secondPlayer) {
    playerPair = [...playerPair, getUpdatedPlayerAfterMatch(secondPlayer, match)]
  }

  return playerPair;
}

/**
 * Shuffles and flattens the list of player idxs. The result is the players array with all match points sorted.
 * 
 * @param chunkedPlayerIdxs Result from chunkSortedArrayByMatchPoints
 * @returns Flattened array with shuffled player idxs.
 */
export const shuffleAndFlattenChunkedPlayersByMatchPoints = (chunkedPlayerIdxs: number[][]): number[] => {
  return chunkedPlayerIdxs.reduce((acc, chunk) => {
    return [...acc, ...shuffle(chunk)]
  }, [])
};

/**
 * Sorts the players array by match points, and scrambles them.
 * Used to sort an unsorted array right after match results are applied.
 * Pairings are just grabbing the players list from the top down.
 * 
 * @param players 
 * @returns 
 */
export const sortPlayersByMatchPoints = (players: Player[]): Player[] => {
  // Players who aren't dropped. Don't pair players who are dropped.
  players = getActivePlayers(players);
  // Sort the players top to bottom by match points.
  players.sort((a, b) => b.matchPoints - a.matchPoints);
  // Get players array chunked by match points.
  const chunkedPlayerIdxs: number[][] = chunkSortedArrayByMatchPoints(players);
  // Shuffle and flatten player idxs
  const shuffledAndFlattenedPlayerIdxs = shuffleAndFlattenChunkedPlayersByMatchPoints(chunkedPlayerIdxs);

  const result = shuffledAndFlattenedPlayerIdxs.map((playerIdx) => players[playerIdx]);
  return result;
};

/**
 * Right after the match is complete and match results are final, applies these match results to the players.
 * @param matchResults 
 * @param players 
 */
export const applyMatchResultsToPlayers = (matches: Match[], players: Player[]): Player[] => {
  return matches.reduce((acc: Player[], match: Match) => [...acc, ...getUpdatedPlayerPairAfterMatch(match, players)], []);
}