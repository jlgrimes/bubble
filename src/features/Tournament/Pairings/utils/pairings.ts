import type { Match, Pairing } from "../types";
import type { Player } from "../../Player/types/Player";
import { applyMatchResultsToPlayers, sortPlayersByMatchPoints } from "../../Player/utils/player";
import { shuffle } from "../../../../helpers/shuffle";

/**
 * Gets pairings from a list of players. Top-down.
 * 
 * @param players Top-down sorted list of players.
 * @returns Pairings list.
 */
export const getPairings = (players: Player[]): Pairing[] => {
  let pairings: Pairing[] = [];
  for (let playerIdx = 0; playerIdx < players.length; playerIdx += 2) {
    if (playerIdx === players.length - 1) {
      pairings.push({ playerIds: [players[playerIdx].id] });
    } else {
      pairings.push({ playerIds: [players[playerIdx].id, players[playerIdx + 1].id] });
    }
  }

  return pairings;
}

/**
 * Gets initial round pairings as an array of matches.
 * 
 * @param players Players list.
 * @returns 
 */
export const getInitialRoundPairings = (players: Player[]): Pairing[] => {
  // Initial round doesn't need to organize by match points.
  const sortedPlayers = shuffle(players);

  return getPairings(sortedPlayers);
};

/**
 * Gets next round pairings as an array of pairings.
 * 
 * @param players Stale players list from state.
 * @returns 
 */
 export const getNextRoundPairings = (players: Player[], matches: Match[]): Pairing[] => {
  // Applies match results to the players list.
  const updatedPlayers = applyMatchResultsToPlayers(matches, players);
  // Sorts the players by match points and scrambles within match point tiers.
  const sortedPlayers = sortPlayersByMatchPoints(updatedPlayers);

  return getPairings(sortedPlayers);
 }