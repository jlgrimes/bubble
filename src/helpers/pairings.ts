import { Match, Pairing } from "../types/Match";
import { Player } from "../types/Player";
import { applyMatchResultsToPlayers, sortPlayersByMatchPoints } from "./player";

/**
 * Gets pairings from a list of players. Top-down.
 * 
 * @param players Top-down sorted list of players.
 * @returns Pairings list.
 */
export const getPairings = (players: Player[]): Pairing[] => {
  let pairings: Pairing[] = [];
  for (let playerIdx = 0; playerIdx < players.length; playerIdx += 2) {
    if (playerIdx === players.length - 1) {
      pairings.push({ playerIds: [players[playerIdx].id] });
    } else {
      pairings.push({ playerIds: [players[playerIdx].id, players[playerIdx + 1].id] });
    }
  }

  return pairings;
}

/**
 * Gets next round pairings as an array of matches.
 * 
 * @param players Stale players list from state
 * @returns 
 */
 export const getNextRoundPairings = (players: Player[], matches: Match[]): Pairing[] => {
  // Applies match results to the players list.
  const updatedPlayers = applyMatchResultsToPlayers(matches, players);
  // Sorts the players by match points and scrambles within match point tiers.
  const sortedPlayers = sortPlayersByMatchPoints(updatedPlayers);

  return getPairings(sortedPlayers);
 }