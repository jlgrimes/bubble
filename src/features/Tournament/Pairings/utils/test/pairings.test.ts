import { getNextRoundPairings, getPairings } from "../pairings";
import { SAMPLE_MATCH_RESULTS, SAMPLE_SORTED_PLAYER_LIST } from "../../../../../helpers/testConstants";

describe('pairings utils', () => {
  describe('getPairings', () => {
    it('should properly get pairings', () => {
      expect(getPairings(SAMPLE_SORTED_PLAYER_LIST)).toMatchSnapshot();
    });

    it('should pair last person as a bye', () => {
      expect(getPairings(SAMPLE_SORTED_PLAYER_LIST.slice(0, 3))).toMatchSnapshot();
    });
  });

  describe('getNextRoundPairings', () => {
    it('should get next round pairings', () => {
      expect(getNextRoundPairings(SAMPLE_SORTED_PLAYER_LIST, SAMPLE_MATCH_RESULTS)).toMatchSnapshot();
    });
  });
});