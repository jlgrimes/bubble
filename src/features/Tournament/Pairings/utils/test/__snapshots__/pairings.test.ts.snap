// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pairings utils getNextRoundPairings should get next round pairings 1`] = `
Array [
  Object {
    "playerIds": Array [
      "0",
      "2",
    ],
  },
  Object {
    "playerIds": Array [
      "2",
      "3",
    ],
  },
]
`;

exports[`pairings utils getPairings should pair last person as a bye 1`] = `
Array [
  Object {
    "playerIds": Array [
      "0",
      "1",
    ],
  },
  Object {
    "playerIds": Array [
      "2",
    ],
  },
]
`;

exports[`pairings utils getPairings should properly get pairings 1`] = `
Array [
  Object {
    "playerIds": Array [
      "0",
      "1",
    ],
  },
  Object {
    "playerIds": Array [
      "2",
      "3",
    ],
  },
]
`;
